### PR TITLE
[RFC][WIP] feat: implement versioning

### PR DESCRIPTION
**Pull request by harry-hov**
This is rough and partial implementation of the proposal that I mentioned [here](https://github.com/gnolang/gno/issues/694#issuecomment-1904296382)

### Changes/Status

- [x] gno.mod is mandatory to publish pkg/realm(s)
- [x] gno.mod's `module` directive takes version now
    e.g: 
    ```go.mod
    module gno.land/p/{handle}/foo v1.2.3
    ```
- [x] pkg/realm(s) are versioned (IOW, multiple version of pkgs/realms can exist with same import path)
- [x] `gnokey maketx addpkg` command takes pkg path and version from gno.mod file
- [x] `gnokey maketx call` command takes `-pkgPath` in the format `gno.land/p/{handle}/foo@v1.2.3`
- [ ] disable previous version of realm when someone publishes new version
- [ ] realm state migration

### Testing
to test, run ([versioning.txtar](https://github.com/harry-hov/gno/blob/hariom/versioning/gno.land/cmd/gnoland/testdata/versioning.txtar)): 
```sh
$ cd gno.land/cmd/gnoland
$ go test -v . -run TestTestdata/versioning
```

<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

File: gno.land/cmd/gnoland/testdata/versioning.txtar [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# test package versioning
```
Comment (harry-hov) [inline gno.land/cmd/gnoland/testdata/versioning.txtar:1]: to test: 
```sh
$ cd gno.land/cmd/gnoland
$ go test -v . -run TestTestdata/versioning
```

File: gno.land/cmd/gnoland/testdata/versioning.txtar [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+gnokey maketx addpkg -pkgdir $WORK/helloworld/v2 -pkgpath gno.land/p/helloworld/hello -gas-fee 1000000ugnot -gas-wanted 2000000 -broadcast -chainid=tendermint_test test1
+
+## add `foo` realm; imports pkg `hello` v1
+gnokey maketx addpkg -pkgdir $WORK/foo -pkgpath gno.land/r/temp/foo -gas-fee 1000000ugnot -gas-wanted 2000000 -broadcast -chainid=tendermint_test test1
+
+## add `bar` realm; imports pkg `hello` v2
+gnokey maketx addpkg -pkgdir $WORK/bar -pkgpath gno.land/r/temp/bar -gas-fee 1000000ugnot -gas-wanted 2000000 -broadcast -chainid=tendermint_test test1
+
+## execute Render from `foo` realm
+gnokey maketx call -pkgpath gno.land/r/temp/foo -pkgversion v0.0.1 -func Render -gas-fee 1000000ugnot -gas-wanted 2000000 -args '' -broadcast -chainid=tendermint_test test1
```
Comment (moul) [inline gno.land/cmd/gnoland/testdata/versioning.txtar:18]:
```suggestion
gnokey maketx call -pkgpath gno.land/r/temp/foo@v0.0.1 -func Render -gas-fee 1000000ugnot -gas-wanted 2000000 -args '' -broadcast -chainid=tendermint_test test1
```
If we use explicit versioning, I suggest something like this.

Comment (moul) [reply gno.land/cmd/gnoland/testdata/versioning.txtar:18]:
```suggestion
gnokey maketx call -pkgpath gno.land/r/temp/foo@v1 -func Render -gas-fee 1000000ugnot -gas-wanted 2000000 -args '' -broadcast -chainid=tendermint_test test1
```

Or, even better, with a shorter version.

File: gno.land/pkg/gnoclient/client_txs.go [hunk 3]
```diff
@@ -65,11 +70,12 @@ func (c *Client) Call(cfg CallCfg) (*ctypes.ResultBroadcastTxCommit, error) {
...
 	msg := vm.MsgCall{
-		Caller:  caller,
-		Send:    sendCoins,
-		PkgPath: pkgPath,
-		Func:    funcName,
-		Args:    args,
+		Caller:     caller,
+		Send:       sendCoins,
+		PkgPath:    pkgPath,
+		PkgVersion: pkgVersion,
```
Comment (moul) [inline gno.land/pkg/gnoclient/client_txs.go:76]:
```suggestion
		PkgPath:    pkgPath + pkgVersion,
```

File: gno.land/cmd/gnoland/testdata/versioning.txtar (hunk 2 above)
Comment (harry-hov) [reply gno.land/cmd/gnoland/testdata/versioning.txtar]: Implemented the first suggestion. As it keeps the versioning similar to Golang/go.mod.

**Comment by zivkovicmilos** [comment]
@harry-hov 

Can you please resolve the merge conflicts in this PR? 🙏

File: gnovm/cmd/gno/mod.go [hunk 4]
```diff
@@ -204,13 +206,70 @@ func execModInit(args []string) error {
...
+			}
+			if pkgName != pn {
+				return fmt.Errorf("package name mismatch: [%q] and [%q]", pkgName, pn)
+			}
+		}
+		if pkgName == "" {
+			return errors.New("cannot determine package name")
+		}
+		modPath = string(pkgName)
+	}
```
Comment (ajnavarro) [inline gnovm/cmd/gno/mod.go:261]: extract this to a function. It will make the code more readable.

File: gnovm/pkg/gnomod/gnomod_test.go [hunk 5]
```diff
@@ -1,151 +1,151 @@
...
-					`module gno.land/p/demo/foo`,
-				},
-			},
-			errShouldContain: "gno.mod file already exists",
-		},
-	} {
-		tc := tc
-		t.Run(tc.desc, func(t *testing.T) {
-			t.Parallel()
+// 	for _, tc := range []struct {
```
Comment (ajnavarro) [inline gnovm/pkg/gnomod/gnomod_test.go:16]: leftover?

File: gno.land/pkg/keyscli/run.go [hunk 6]
```diff
@@ -98,16 +99,25 @@ func execMakeRun(cfg *MakeRunCfg, args []string, cmdio commands.IO) error {
...
-			memPkg.Files = []*std.MemFile{
-				{
-					Name: info.Name(),
-					Body: string(b),
+
+			gm := gno.ParseMemMod(filepath.Dir(sourcePath))
+			// TODO(hariom): Remove
+			// if err != nil {
+			// 	return fmt.Errorf("error parse gno.mod at: %q", sourcePath)
+			// }
```
Comment (ajnavarro) [inline gno.land/pkg/keyscli/run.go:107]: leftover?

File: gnovm/cmd/gno/mod.go [hunk 7]
```diff
@@ -204,13 +206,70 @@ func execModInit(args []string) error {
...
+	if err := module.CheckImportPath(modPath); err != nil {
+		return err
+	}
+
+	modfile := new(gnomod.File)
+	modfile.AddModuleStmt(modPath)
+	modfile.Write(filepath.Join(rootDir, "gno.mod"))
```
Comment (ajnavarro) [inline gnovm/cmd/gno/mod.go:268]: unhandled errors...

**Review by ajnavarro — CHANGES_REQUESTED** [review]
If we want to keep following semver, we have to change the first release to v0.1.0:

https://semver.org/spec/v2.0.0.html#how-should-i-deal-with-revisions-in-the-0yz-initial-development-phase

File: gnovm/cmd/gno/mod.go (hunk 4 above)
Comment (harry-hov) [reply gnovm/cmd/gno/mod.go:261]: makes sense. 

It's not introduced in this PR, I just moved this function to a new file in this PR. 
I will open another PR soon to divide it into smaller parts.

File: gno.land/pkg/keyscli/run.go (hunk 6 above)
Comment (harry-hov) [reply gno.land/pkg/keyscli/run.go:107]: yeah. I will remove it in the cleanup once I fix all the failing tests.

File: gnovm/pkg/gnomod/gnomod_test.go (hunk 5 above)
Comment (harry-hov) [reply gnovm/pkg/gnomod/gnomod_test.go:16]: yet another failing test. I will uncomment once I fix the other failing ones.

File: gnovm/cmd/gno/mod.go (hunk 7 above)
Comment (harry-hov) [reply gnovm/cmd/gno/mod.go:268]: > It's not introduced in this PR, I just moved this function to a new file in this PR.
I will open another PR soon to divide it into smaller parts.

I will also fix it in the same PR

**Comment by harry-hov** [comment]
> If we want to keep following semver, we have to change the first module release to v0.1.0:
>
> https://semver.org/spec/v2.0.0.html#how-should-i-deal-with-revisions-in-the-0yz-initial-development-phase

I agree. changed it to `v0.1.0`  🙌 

Thanks @ajnavarro

File: gnovm/pkg/gnolang/preprocess.go [hunk 8]
```diff
@@ -3031,11 +3035,32 @@ func tryPredefine(store Store, last BlockNode, d Decl) (un Name) {
...
+		pn := packageOf(last)
+		var version string
+		if strings.HasPrefix(d.PkgPath, "gno.land") {
+			for _, req := range pn.ModFile.Require {
+				if req.Path == d.PkgPath {
+					version = req.Version
+					break
+				}
+			}
+			// TODO(hariom): uncomment?
```
Comment (zivkovicmilos) [inline gnovm/pkg/gnolang/preprocess.go:3047]: Leftover?

File: gnovm/pkg/gnolang/preprocess.go [hunk 9]
```diff
@@ -3031,11 +3035,32 @@ func tryPredefine(store Store, last BlockNode, d Decl) (un Name) {
...
+			// - If package doen't contain gno.mod
+			// - Testing _filetest. Currently imports are covered by gno.mod
+			// if version == "" {
+			// 	panic(fmt.Sprintf(
+			// 		"cannot get version for package %s",
+			// 		d.PkgPath))
+			// }
+		}
+
+		// TODO(hariom): create helper
```
Comment (zivkovicmilos) [inline gnovm/pkg/gnolang/preprocess.go:3058]: Leftover?

File: gnovm/pkg/gnolang/preprocess.go [hunk 10]
```diff
@@ -3031,11 +3035,32 @@ func tryPredefine(store Store, last BlockNode, d Decl) (un Name) {
...
+		pv := store.GetPackage(d.PkgPath, version, true) // TODO(hariom): temp; fix
```
Comment (zivkovicmilos) [inline gnovm/pkg/gnolang/preprocess.go:3059]: Leftover?

File: gnovm/pkg/gnolang/preprocess.go [hunk 11]
```diff
@@ -3614,9 +3642,10 @@ func SaveBlockNodes(store Store, fn *FileNode) {
 	// First, get the package and file names.
 	pn := packageOf(fn)
 	store.SetBlockNode(pn)
-	pkgPath := pn.PkgPath
+	pkgPath := pn.ModFile.Path
+	pkgVersion := pn.ModFile.Version
 	fileName := string(fn.Name)
-	if pkgPath == "" || fileName == "" {
+	if pkgPath == "" || fileName == "" { //  TODO(hariom): assert empty pkgVersion if not stdlib
```
Comment (zivkovicmilos) [inline gnovm/pkg/gnolang/preprocess.go:3648]: Leftover?

File: gnovm/pkg/gnolang/types.go [hunk 12]
```diff
@@ -1434,7 +1434,7 @@ func (dt *DeclaredType) TypeID() TypeID {
 	return dt.typeid
 }
 
-func DeclaredTypeID(pkgPath string, name Name) TypeID {
+func DeclaredTypeID(pkgPath string, name Name) TypeID { // TODO(hariom): include version?
```
Comment (zivkovicmilos) [inline gnovm/pkg/gnolang/types.go:1437]: Leftover?

File: gnovm/pkg/gnolang/utils.go [hunk 13]
```diff
@@ -19,3 +25,27 @@ func endsWith(item string, suffixes []string) bool {
 	}
 	return false
 }
+
+// TODO(hariom): move to better place
```
Comment (zivkovicmilos) [inline gnovm/pkg/gnolang/utils.go:29]: Leftover?

File: gnovm/pkg/gnolang/utils.go [hunk 14]
```diff
@@ -19,3 +25,27 @@ func endsWith(item string, suffixes []string) bool {
...
+func ParseMemMod(dir string) *std.MemMod {
```
Comment (zivkovicmilos) [inline gnovm/pkg/gnolang/utils.go:30]: Why not have this method return an error?

File: gnovm/pkg/gnolang/utils.go [hunk 15]
```diff
@@ -19,3 +25,27 @@ func endsWith(item string, suffixes []string) bool {
...
+	var memMod std.MemMod
```
Comment (zivkovicmilos) [inline gnovm/pkg/gnolang/utils.go:31]: Why declare here, when you're actually setting the object at the end of the method?

File: gnovm/pkg/gnolang/uverse.go [hunk 16]
```diff
@@ -64,6 +66,7 @@ var (
 )
 
 const uversePkgPath = ".uverse"
+const uversePkgVersion = "v0.1.0"
```
Comment (zivkovicmilos) [inline gnovm/pkg/gnolang/uverse.go:70]: Nitpick: you can combine this in a common `const`

File: gnovm/pkg/gnolang/machine.go [hunk 17]
```diff
@@ -55,18 +56,20 @@ type Machine struct {
 //
 // Like for [NewMachineWithOptions], Machines initialized through this
 // constructor must be finalized with [Machine.Release].
-func NewMachine(pkgPath string, store Store) *Machine {
+func NewMachine(pkgPath, pkgVersion string, store Store) *Machine {
```
Comment (zivkovicmilos) [inline gnovm/pkg/gnolang/machine.go:59]: What do you think about making `NewMachine` take in the `Store`, and have the other params as options?
As far as I can see, there is only one instance where `pkgVersion != """`

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
I like the idea behind this PR, but I don't think it has been executed correctly, yet 🙁 

I am _really_ not a fan of interacting with Realms (calling their methods...) by specifying their **version number**, alongside the path.
The path is clear, and the version should be resolved from the path when calling Realm methods (latest). 

There's a lot to unpack here, I think we need to collectively decide the API for versioning before we continue

cc @moul @thehowl

**Comment by harry-hov** [comment]
Initially, I envisioned this PR to serve as a proof of concept for versioning. I marked it as ready for review to draw attention to this topic and to receive feedback on the idea in general.

As directed by @zivkovicmilos, I am converting this to a draft until I complete the implementation.

File: gnovm/pkg/gnolang/frame.go [hunk 18]
```diff
@@ -39,7 +39,7 @@ func (fr Frame) String() string {
 			fr.NumExprs,
 			fr.NumStmts,
 			fr.NumBlocks,
-			fr.LastPackage.PkgPath,
+			fr.LastPackage.ModFile.Path,
```
Comment (piux2) [inline gnovm/pkg/gnolang/frame.go:42]: We should not implement mod and package versioning at the VM level. The module is a development tool that manages package versioning outside of VM execution.

File: gnovm/pkg/gnolang/op_types.go [hunk 19]
```diff
@@ -142,7 +142,7 @@ func (m *Machine) doOpInterfaceType() {
 	}
 	// push interface type
 	it := &InterfaceType{
-		PkgPath: m.Package.PkgPath,
+		PkgPath: m.Package.ModFile.Path,
```
Comment (piux2) [inline gnovm/pkg/gnolang/op_types.go:145]: We should not implement mod and package versioning at the VM level. The module is a development tool that manages package versioning outside of VM execution.

Same as many other places in this implementation

**Comment by Kouteki** [comment]
@harry-hov is this a breaking change? Trying to determine whether we can do this feature post test4 launch.

**Comment by harry-hov** [comment]
> @harry-hov is this a breaking change? Trying to determine whether we can do this feature post test4 launch.

@Kouteki Yes it's a breaking change. It changes how you publish and interact with gno packages. 

I have no problem with moving this post launch. Regardless, it's on my priority list.

**Comment by thehowl** [comment]
Closing this PR for now; better proposals are welcome.

**Pull request closed without merging**